src/main/java/com/snowflake/kafka/connector/internal/StageFilesProcessor.java

```diff
@@ -0,0 +1,505 @@
+        ProgressRegisterImpl register = new ProgressRegisterImpl(pipeCreation, pipeTelemetry, telemetryService);
+        register.owner = this;
+
+        cleanerExecutor.submit(() -> trackFiles(
+                register,
+                Thread.currentThread(),
+                () -> !Thread.currentThread().isInterrupted(),
+                Thread::sleep));
+
+        return register;
+    }
+
+    /**
+     * trackFiles - actual processor logic with extracted thread management for testing purposes.
+     * @param register - client interface
+     * @param worker - normally - common pool worker thread doing the work; allows providing custom thread for testing
+     * @param canContinue - delegate to Thread.currentThread().isInterrupted(), required for testing
+     * @param awaiter - delegate to Thread.sleep(), required for testing
+     */
+    @VisibleForTesting
+    void trackFiles(ProgressRegisterImpl register, Thread worker, Supplier<Boolean> canContinue, TaskAwaiter awaiter) {
+        runnerThread.set(worker);
+        boolean shouldFetchInitialStageFiles = true;
+        boolean firstRun = true;
+
+        List<String> files = new ArrayList<>();
+        while (canContinue.get()) {
+            try {
+                register.getTelemetry().reportKafkaPartitionUsage(register.getPipeCreation(), false);
+
+                // wait 1 minute
+                awaiter.await(Duration.ofMinutes(1L).toMillis());
+                // do we need to build some initial state based on the remote state?
+                if (shouldFetchInitialStageFiles) {
+                    files.addAll(fetchCurrentStage());
+                    shouldFetchInitialStageFiles = false;
+                }
+                // add all files which might have been collected in the past minute while waiting for data
+                register.files.drainTo(files);
+                // process the files, store the spillover ones for the next cycle. in case of an error - we
+                // will retry processing with the current file set in next iteration (with potentially newly added files)
+                files = nextCheck(files, register, firstRun);
+            } catch (InterruptedException e) {
+                LOGGER.info("Cleaner terminated by an interrupt:\n{}", e.getMessage());
+                worker.interrupt();
+                return;
+            } catch (Exception e) {
+                register.getTelemetry().reportKafkaConnectFatalError(e.getMessage());
+                LOGGER.warn(
+                        "Cleaner encountered an exception {}:\n{}\n{}",
+                        e.getClass(),
+                        e.getMessage(),
+                        e.getStackTrace());
+                shouldFetchInitialStageFiles = true;
+            }
+            firstRun = false;
+        }
+    }
+
+    @Override
+    public void close() {
+        Thread runner = runnerThread.getAndSet(null);
```

Comment:
first caller gets the reference and breaks the cleaner loop. all other will always get null and will not proceed. 
yes, it is ok to interrupt them - that's the best signal for the executing in methods like sleep or doing I/O to actually give up on execution and return control. 
i'm handling the interrupt exception in the trackFiles method, so in such case the interrupted flag would be cleared and thread returns to the pool as it would normally complete its job.